SIGSEGV when reconciling a a helm chart from git with missing dependencies

If someone applies a GitRepository and corresponding HelmRelease that points to a helm chart that is missing dependencies (specifically this one: https://github.com/helxplatform/search-chart/tree/0e35ab7713106421781388eaa5d8daaa275fd30f ), the source-controller will segfault and crashloop.

The reason appears to be that running `helm package` inside that git repo fails with the error:

```
Error: found in Chart.yaml, but missing in charts/ directory: airflow, elasticsearch, redis, tranql
```

But looking through the stack trace, that error gets passed up through the call chain until some point where someone might have forgotten to check the `err`.

```
{"level":"info","ts":"2022-07-05T18:25:26.703Z","logger":"controller.helmrepository","msg":"Starting workers","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"HelmRepository","worker count":2}
{"level":"info","ts":"2022-07-05T18:25:26.703Z","logger":"controller.gitrepository","msg":"Starting workers","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"GitRepository","worker count":2}
{"level":"info","ts":"2022-07-05T18:25:26.703Z","logger":"controller.helmrepository","msg":"Starting workers","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"HelmRepository","worker count":2}
{"level":"info","ts":"2022-07-05T18:25:26.703Z","logger":"controller.helmchart","msg":"Starting workers","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"HelmChart","worker count":2}
{"level":"info","ts":"2022-07-05T18:25:26.703Z","logger":"controller.bucket","msg":"Starting workers","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"Bucket","worker count":2}
{"level":"info","ts":"2022-07-05T18:25:26.845Z","logger":"controller.gitrepository","msg":"no changes since last reconcilation: observed revision 'develop/0e35ab7713106421781388eaa5d8daaa275fd30f'","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"GitRepository","name":"search-repo","namespace":"tcheek","reconcileID":"a4569e79-ccb7-4dc6-aaf1-c23541885138"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x1d5b434]

goroutine 659 [running]:
github.com/fluxcd/source-controller/internal/helm/repository.(*ChartRepository).CacheIndexInMemory(0xc0008df4d0)
    github.com/fluxcd/source-controller/internal/helm/repository/chart_repository.go:351 +0x14
github.com/fluxcd/source-controller/internal/helm/chart.(*DependencyManager).Clear(0xc0008df6e0)
    github.com/fluxcd/source-controller/internal/helm/chart/dependency_manager.go:100 +0xcd
github.com/fluxcd/source-controller/controllers.(*HelmChartReconciler).buildFromTarballArtifact(0xc00004fc80, {0x2a55c58, 0xc0009a00f0}, 0xc000985200, {{0xc000436f00, 0x50}, {0xc
000740000, 0x88}, {0xc000d894a0, 0x30}, ...}, ...)
    github.com/fluxcd/source-controller/controllers/helmchart_controller.go:739 +0xf34
github.com/fluxcd/source-controller/controllers.(*HelmChartReconciler).reconcileSource(0xc00004fc80, {0x2a55c58, 0xc0009a00f0}, 0xc000985200, 0xc00100f9e0)
    github.com/fluxcd/source-controller/controllers/helmchart_controller.go:443 +0x905
github.com/fluxcd/source-controller/controllers.(*HelmChartReconciler).reconcile(0x2ac5360, {0x2a55c58, 0xc0009a00f0}, 0xc000985200, {0xc001c4bbf0, 0x3, 0xc000710ba6})
    github.com/fluxcd/source-controller/controllers/helmchart_controller.go:281 +0x1a4
github.com/fluxcd/source-controller/controllers.(*HelmChartReconciler).Reconcile(0xc00004fc80, {0x2a55c58, 0xc0009a0030}, {{{0xc000710ba6, 0x25d5f60}, {0xc00091a4c8, 0x30}}})
    github.com/fluxcd/source-controller/controllers/helmchart_controller.go:259 +0x5b0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0006c42c0, {0x2a55c58, 0xc000982b40}, {{{0xc000710ba6, 0x25d5f60}, {0xc00091a4c8, 0x41aa670}}})
    sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114 +0x26f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0006c42c0, {0x2a55bb0, 0xc000f1d340}, {0x245dc80, 0xc000156160})
    sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311 +0x33e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0006c42c0, {0x2a55bb0, 0xc000f1d340})
    sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
    sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
    sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:223 +0x357
```

--- a/internal/helm/chart/dependency_manager.go
+++ b/internal/helm/chart/dependency_manager.go
@@ -98,7 +98,9 @@ func NewDependencyManager(opts ...DependencyManagerOption) *DependencyManager {
 func (dm *DependencyManager) Clear() error {
 	var errs []error
 	for _, v := range dm.downloaders {
-		errs = append(errs, v.Clear())
+		if v != nil {
+			errs = append(errs, v.Clear())
+		}
 	}
 	return errors.NewAggregate(errs)
 }
@@ -257,6 +259,10 @@ func (dm *DependencyManager) resolveRepository(url string) (repo repository.Down
 	defer dm.mu.Unlock()
 
 	nUrl := repository.NormalizeURL(url)
+	err = repository.ValidateDepURL(nUrl)
+	if err != nil {
+		return
+	}
 	if _, ok := dm.downloaders[nUrl]; !ok {
 		if dm.getChartDownloaderCallback == nil {
 			err = fmt.Errorf("no chart repository for URL '%s'", nUrl)

--- a/internal/helm/chart/dependency_manager_test.go
+++ b/internal/helm/chart/dependency_manager_test.go
@@ -93,6 +93,7 @@ func TestDependencyManager_Clear(t *testing.T) {
 		},
 		"with credentials":    ociRepoWithCreds,
 		"without credentials": &repository.OCIChartRepository{},
+		"nil downloader":      nil,
 	}
 
 	dm := NewDependencyManager(WithRepositories(downloaders))
@@ -427,6 +428,14 @@ func TestDependencyManager_addRemoteDependency(t *testing.T) {
 				Repository: "https://example.com",
 			},
 			wantErr: "no chart repository for URL",
+		},
+		{
+			name:        "resolve aliased repository error",
+			downloaders: map[string]repository.Downloader{},
+			dep: &helmchart.Dependency{
+				Repository: "@fantastic-charts",
+			},
+			wantErr: "aliased repository dependency is not supported",
 		},
 		{
 			name: "strategic load error",

--- a/internal/helm/repository/utils.go
+++ b/internal/helm/repository/utils.go
@@ -17,9 +17,21 @@ limitations under the License.
 package repository
 
 import (
+	"fmt"
 	"strings"
 
 	helmreg "helm.sh/helm/v3/pkg/registry"
+)
+
+const (
+	alias = "@"
+)
+
+var (
+	// errInvalidDepURL is returned when the dependency URL is not supported
+	errInvalidDepURL = fmt.Errorf("invalid dependency repository URL")
+	// errInvalidAliasedDep is returned when the dependency URL is an alias
+	errInvalidAliasedDep = fmt.Errorf("aliased repository dependency is not supported")
 )
 
 // NormalizeURL normalizes a ChartRepository URL by its scheme.
@@ -34,4 +46,20 @@ func NormalizeURL(repositoryURL string) string {
 
 	return strings.TrimRight(repositoryURL, "/") + "/"
 
+}
+
+// ValidateDepURL returns an error if the given depended repository URL declaration is not supported
+// The reason for this is that the dependency manager will not be able to resolve the alias declaration
+// e.g. repository: "@fantastic-charts"
+func ValidateDepURL(repositoryURL string) error {
+	switch {
+	case strings.HasPrefix(repositoryURL, helmreg.OCIScheme):
+		return nil
+	case strings.HasPrefix(repositoryURL, "https://") || strings.HasPrefix(repositoryURL, "http://"):
+		return nil
+	case strings.HasPrefix(repositoryURL, alias):
+		return fmt.Errorf("%w: %s", errInvalidAliasedDep, repositoryURL)
+	default:
+		return fmt.Errorf("%w: %s", errInvalidDepURL, repositoryURL)
+	}
 }

